### PR TITLE
Android: add option to disable TLS validation and allow self-signed certificates for WebDAV/NextCloud

### PR DESCRIPTION
**Pull request by roman-r-m**
The change consists of 2 parts:
* For the react-native built-in fetch it uses a hook that RN provides (`NetworkingModule.setCustomClientBuilder`)
* For rn-fetch-blob sets `trusty` option which does more or less the same thing under the hood

Tested, works for me. ~~One thing that is still missing is "Check Synchronization Configuration" - at the moment you must save for the change to be applied.~~

Resolves: https://github.com/laurent22/joplin/issues/4439 & https://github.com/laurent22/joplin/issues/4479 & https://github.com/laurent22/joplin/issues/4487 & https://github.com/laurent22/joplin/issues/4644

### CODE REVIEW

**Comment by roman-r-m** [comment]
Laurent, do we want to also allow this for Joplin server? I think it makes sense.

**Comment by roman-r-m** [comment]
One more way in which it can be improved is instead of disabling validation altogether, allow users to set their own certs and use them in addition to built-in system certificates.
I think the code for this should be roughly the same, just a bit more complicated as it'll need to copy certs somewhere and pass them to native code.

**Comment by mu88** [comment]
Do you have a rough idea when this feature will ship?

**Comment by roman-r-m** [comment]
> Do you have a rough idea when this feature will ship?

This is outside of my control, sorry

**Comment by laurent22** [comment]
Thanks for looking into this @roman-r-m but I don't really know how to review this actually since I don't know the Android internal code much.

React Native network code is buggy and unreliable in general and my concern is that if we tweak it, it might break either in some edge case, or later when RN is upgraded. We also don't have anything to unit test the Java code (and I think it would be out of scope to add this).

So I'm wondering, how can we know for sure that the extension is not going to break something? And how resilient it is to RN network code changing from update to update?

Actually is it possible to make sure the extension is active only when the "ignore ssl errors" option is on? Like it doesn't go through this code path at all if it's disabled?

**Comment by roman-r-m** [comment]
> React Native network code is buggy and unreliable in general and my concern is that if we tweak it, it might break either in some edge case, or later when RN is upgraded. We also don't have anything to unit test the Java code (and I think it would be out of scope to add this).

With regards to future RN upgrades, I'd expect that if anything changes in the API used here it just won't compile. There's still a chance of more subtle bugs, of course, but not sure if it's at all possible to avoid those.

Give how often this is requested I feel the benefits may outweigh any potential issues.

> Actually is it possible to make sure the extension is active only when the "ignore ssl errors" option is on? Like it doesn't go through this code path at all if it's disabled?

That's not possible because if the user enables the validation again after disabling it, we need to undo the changes. Best I can do is to not call this native module on startup if the setting is not set (this is now done)

File: packages/app-mobile/components/screens/config.js [hunk 1]
```diff
@@ -38,7 +39,13 @@ class ConfigScreenComponent extends BaseScreenComponent {
 		shared.init(this);
 
 		this.checkSyncConfig_ = async () => {
-			await shared.checkSyncConfig(this, this.state.settings);
+			// to ignore TLS erros we need to chage the global state of the app, if the check fails we need to restore the original state
+			// this call sets the new value and returns the previous one which we can use later to revert the change
+			const prevIgnoreTlsErrors = await NativeModules.SslModule.setIgnoreTlsErrors(this.state.settings['net.ignoreTlsErrors']);
```
Comment (roman-r-m) [inline packages/app-mobile/components/screens/config.js:44]: just realized this is broken on iOS, will fix now

Comment (roman-r-m) [reply packages/app-mobile/components/screens/config.js]: fixed now

**Comment by laurent22** [comment]
Thanks for clarifying @roman-r-m. I think it makes sense then, so let's merge.
